binary-option/program/src/processor.rs

```diff
@@ -0,0 +1,725 @@
+    binary_option.escrow_account_pubkey = *escrow_account_info.key;
+    binary_option.owner = *update_authority_info.key;
+    binary_option.serialize(&mut *binary_option_account_info.data.borrow_mut())?;
+
+    Ok(())
+}
+
+pub fn process_trade(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    size: u64,
+    buy_price: u64,
+    sell_price: u64,
+) -> ProgramResult {
+    msg!("Trade");
+    let account_info_iter = &mut accounts.iter();
+    let binary_option_account_info = next_account_info(account_info_iter)?;
+    let escrow_account_info = next_account_info(account_info_iter)?;
+    let long_token_mint_info = next_account_info(account_info_iter)?;
+    let short_token_mint_info = next_account_info(account_info_iter)?;
+    let buyer_info = next_account_info(account_info_iter)?;
+    let seller_info = next_account_info(account_info_iter)?;
+    let buyer_account_info = next_account_info(account_info_iter)?;
+    let seller_account_info = next_account_info(account_info_iter)?;
+    let buyer_long_token_account_info = next_account_info(account_info_iter)?;
+    let buyer_short_token_account_info = next_account_info(account_info_iter)?;
+    let seller_long_token_account_info = next_account_info(account_info_iter)?;
+    let seller_short_token_account_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+    let token_program_info = next_account_info(account_info_iter)?;
+
+    // Unpack accounts
+    let long_token_mint: Mint = assert_initialized(long_token_mint_info)?;
+    let short_token_mint: Mint = assert_initialized(short_token_mint_info)?;
+    let buyer_long_token_account: Account = assert_initialized(buyer_long_token_account_info)?;
+    let buyer_short_token_account: Account = assert_initialized(buyer_short_token_account_info)?;
+    let seller_long_token_account: Account = assert_initialized(seller_long_token_account_info)?;
+    let seller_short_token_account: Account = assert_initialized(seller_short_token_account_info)?;
+    let buyer_account: Account = assert_initialized(buyer_account_info)?;
+    let seller_account: Account = assert_initialized(seller_account_info)?;
+    let mut binary_option = BinaryOption::try_from_slice(&binary_option_account_info.data.borrow_mut())?;
+
+    // Get program derived address for escrow
+    let (authority_key, bump_seed) = Pubkey::find_program_address(
+        &[
+            long_token_mint_info.key.as_ref(),
+            short_token_mint_info.key.as_ref(),
+            token_program_info.key.as_ref(),
+            program_id.as_ref(),
+        ],
+        program_id,
+    );
+    let seeds = &[
+        long_token_mint_info.key.as_ref(),
+        short_token_mint_info.key.as_ref(),
+        token_program_info.key.as_ref(),
+        program_id.as_ref(),
+        &[bump_seed],
+    ];
+
+    // Validate data
+    if buy_price + sell_price != u64::pow(10, binary_option.decimals as u32) {
```

Comment:
I think decimal should be at least 2. Can add that in here

Reply:
Please do -- even if `decimal = 10`, with a non-checked add it's possible to overflow with the right inputs

Reply:
I guess you would need an overflow to happen here for this to be dangerous, but I'm pretty sure the transaction should just still fail because the user with the massive funds shouldn't have a wallet with that quantity of tokens. I can probably add a sanity check here just in case

Reply:
Looking into this some more... since `buy_price` and `sell_price` come from the outside, it's very easy to overflow. so you could probably create a transaction that liquidates the escrow account during a redemption trade.